src/pool/mempool.rs

```diff
@@ -21,95 +33,146 @@ pub type KakarotPool<Client> = Pool<Validator<Client>, TransactionOrdering, Noop
-                return Self::default();
+                return Ok(Self::default());
             }
         };
 
         // Extract the account addresses from the JSON array
         if let Some(array) = json.as_array() {
             for item in array {
-                if let Some(address) = item.as_str() {
-                    accounts.insert(Felt::from_hex_unchecked(address));
+                if let Some(account_address) = item.as_str() {
+                    let felt_address = Felt::from_hex_unchecked(account_address);
+
+                    let starknet_block_id = eth_provider
+                        .to_starknet_block_id(Some(BlockId::default()))
+                        .await
+                        .map_err(|e| eyre::eyre!("Error converting block ID: {:?}", e))?;
+
+                    // Query the initial account_nonce for the account from the provider
+                    accounts.insert(
+                        felt_address,
+                        eth_provider
+                            .starknet_provider()
+                            .get_nonce(starknet_block_id, felt_address)
+                            .await
+                            .unwrap_or_default(),
+                    );
                 }
             }
         }
 
-        Self { accounts }
+        Ok(Self { accounts: Arc::new(Mutex::new(accounts)) })
     }
 
-    pub fn start<SP>(&self, rt_handle: &Handle, eth_client: &'static EthClient<SP>)
+    /// Start the account manager task.
+    #[allow(clippy::significant_drop_tightening)]
+    pub fn start<SP>(&self, rt_handle: &Handle, eth_client: Arc<EthClient<SP>>)
     where
         SP: starknet::providers::Provider + Send + Sync + Clone + 'static,
     {
         let accounts = self.accounts.clone();
 
         rt_handle.spawn(async move {
             loop {
-                for address in &accounts {
-                    Self::process_transaction(address, eth_client);
+                let result: eyre::Result<()> = {
+                    let mut accounts = accounts.lock().await;
+
+                    // Iterate over the accounts
+                    for (account_address, account_nonce) in accounts.iter_mut() {
+                        match Self::get_balance(account_address, &eth_client).await {
+                            Ok(balance) => {
+                                if balance > U256::from(u128::pow(10, 18)) {
+                                    Self::process_transaction(account_address, account_nonce, &eth_client);
+                                }
+                            }
+                            Err(e) => {
+                                tracing::error!(
+                                    "Error getting balance for account_address {:?}: {:?}",
+                                    account_address,
+                                    e
+                                );
+                            }
+                        }
+                    }
```

Comment:
can we replace this by something like
```
let balance = Self::get_balance(account_address, &eth_client).await.inspect_err(|err| tracing::error!(...))?;
if balance > ... {
   ...
}
```

Reply:
I think this is a little bit trickier than this because we can't directly propagate the error like this inside the async loop. So I've just proposed another approach quite similar where we store the error inside an iter_err to return at the end. Let me know what you think about it. I think it's the easier and common approach